lls/main.c

```diff
@@ -16,11 +16,218 @@
+		 * bonded interface.
+		 *
+		 * See: http://dpdk.org/doc/guides/prog_guide/link_bonding_poll_mode_drv_lib.html#configuration
+		 */
+		if (!is_broadcast_ether_addr(&eth_hdr->d_addr) &&
+		    !is_same_ether_addr(&eth_hdr->d_addr, &iface->eth_addr))
+			goto free_buf;
+
+		/* TODO Add support for ARP and ND packets. */
+		switch (rte_be_to_cpu_16(eth_hdr->ether_type)) {
+		default:
+			RTE_LOG(ERR, GATEKEEPER, "lls: %s interface should not be seeing a packet with EtherType 0x%04hx\n",
+				iface->name,
+				rte_be_to_cpu_16(eth_hdr->ether_type));
+			goto free_buf;
+		}
+free_buf:
+		rte_pktmbuf_free(bufs[i]);
+	}
+}
+
+static int
+lls_proc(void *arg)
+{
+	struct lls_config *lls_conf = (struct lls_config *)arg;
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"lls: the LLS block is running at lcore = %u\n",
+		lls_conf->lcore_id);
+
+	/* Wait for network devices to start. */
+	while (lls_conf->net->configuring)
+		;
+
+	/* TODO Add ARP and ND filters. */
+
+	while (likely(!exiting)) {
+		/* Read in packets on front and back interfaces. */
+		process_pkts(lls_conf, &lls_conf->net->front,
+			lls_conf->rx_queue_front, lls_conf->tx_queue_front);
+		if (lls_conf->net->back_iface_enabled)
+			process_pkts(lls_conf, &lls_conf->net->back,
+				lls_conf->rx_queue_back,
+				lls_conf->tx_queue_back);
+
+		/* Process any updates or requests. */
+		if (likely(lls_cache_writer(lls_conf) == 0)) {
+			/*
+			 * If there are no requests to go through, then do a
+			 * scan of the cache (if enough time has passed).
+			 *
+			 * XXX In theory, many new LLS changes could starve
+			 * the ability to scan, but this will not likely
+			 * happen. In fact, we may want to reduce the amount
+			 * of times this is called, since reading the HPET
+			 * timer is inefficient. See the timer application.
+			 */
+			rte_timer_manage();
+		}
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"lls: the LLS block at lcore = %u is exiting\n",
+		lls_conf->lcore_id);
+
+	return cleanup_lls();
+}
+
+static int
+assign_lls_queue_ids(struct lls_config *lls_conf)
+{
+	int ret = get_queue_id(&lls_conf->net->front, QUEUE_TYPE_RX,
+		lls_conf->lcore_id);
+	if (ret < 0)
+		return ret;
+	lls_conf->rx_queue_front = (uint16_t)ret;
+
+	ret = get_queue_id(&lls_conf->net->front, QUEUE_TYPE_TX,
+		lls_conf->lcore_id);
+	if (ret < 0)
+		return ret;
+	lls_conf->tx_queue_front = (uint16_t)ret;
+
+	if (lls_conf->net->back_iface_enabled) {
+		ret = get_queue_id(&lls_conf->net->back, QUEUE_TYPE_RX,
+			lls_conf->lcore_id);
+		if (ret < 0)
+			return ret;
+		lls_conf->rx_queue_back = (uint16_t)ret;
+
+		ret = get_queue_id(&lls_conf->net->back, QUEUE_TYPE_TX,
+			lls_conf->lcore_id);
+		if (ret < 0)
+			return ret;
+		lls_conf->tx_queue_back = (uint16_t)ret;
+	}
+
+	return 0;
+}
 
 int
-run_lls(void)
+run_lls(struct net_config *net_conf, struct lls_config *lls_conf)
 {
-	/* TODO Initialize and run Link Layer Support functional block. */
+	int ret;
+
+	if (net_conf == NULL) {
```

Comment:
`run_lls()` is called from Lua, so we should play unusually safe here and test `lls_conf` too.